inferno-ml-remote/test/Main.hs

```diff
@@ -0,0 +1,164 @@
+      inferenceClient = runInference api
+
+      runClient :: Int -> ClientM a -> IO (Either ClientError a)
+      runClient port = flip runClientM $ mkEnv port
+
+      mkEvalTest :: FilePath -> EvalResult -> SpecWith (Arg (Int -> IO ()))
+      mkEvalTest fp expected =
+        Hspec.it ("evaluates " <> takeFileName fp) $ \port ->
+          runClient port (inferenceClient =<< readScript fp)
+            >>= either
+              (Hspec.expectationFailure . show)
+              (`Hspec.shouldBe` expected)
+
+  Hspec.around (withTestAppAndEnv mkCacheOption) $ do
+    Hspec.describe "POST /inference" $ do
+      mkEvalTest "../inferno-ml/test/test.inferno" "Tensor Int64 []  262144"
+
+      mkEvalTest "../inferno-ml/test/test-cpu.inferno" "Tensor Float []  8.5899e9"
+
+      mkEvalTest "../inferno-ml/test/mnist.inferno" "()"
+
+  -- This tests caching and using the model with the filesystem compression
+  -- option, the one above uses non-compressed paths
+  Hspec.around (withTestAppAndEnv mkCompressedCacheOption) $ do
+    Hspec.describe "POST /inference (compressed)" $ do
+      mkEvalTest "../inferno-ml/test/mnist.inferno" "()"
+
+collectModelsSpec :: Spec
+collectModelsSpec = Hspec.describe "collectModelNames" $ do
+  Hspec.it "extracts models from script" $ do
+    mkAstTest "../inferno-ml/test/mnist.inferno" $
+      (`Hspec.shouldBe` ["mnist.ts.pt"]) . collectModelNames
+
+  Hspec.it "does not extract superfluous models" $ do
+    mkAstTest "./test/contrived5.inferno" $
+      (`Hspec.shouldBe` []) . collectModelNames
+
+  Hspec.it "extracts multiple models from script" $ do
+    mkAstTest "./test/contrived.inferno" $
+      (`Hspec.shouldBe` ["x.ts.pt", "y.ts.pt"]) . collectModelNames
+
+  Hspec.it "extracts a single model with non-literal name" $ do
+    mkAstTest "./test/contrived4.inferno" $
+      (`Hspec.shouldBe` ["x.ts.pt"]) . collectModelNames
+
+  Hspec.it "extracts multiple models with non-literal names" $ do
+    mkAstTest "./test/contrived2.inferno" $
+      (`Hspec.shouldBe` ["x.ts.pt", "y.ts.pt"]) . collectModelNames
+
+  Hspec.it "extracts multiple models with literal and non-literal names" $ do
+    mkAstTest "./test/contrived2.inferno" $
+      (`Hspec.shouldBe` ["x.ts.pt", "y.ts.pt"]) . collectModelNames
+  where
+    mkAstTest :: FilePath -> (Expr (Maybe VCObjectHash) SourcePos -> IO ()) -> IO ()
+    mkAstTest fp f =
+      either (Hspec.expectationFailure . displayException) f
+        =<< astFromScript fp
+
+    astFromScript ::
+      FilePath -> IO (Either SomeInfernoError (Expr (Maybe VCObjectHash) SourcePos))
+    astFromScript = fmap (fmap snd . mkFinalAst) . readScript
+
+cacheModelsSpec :: Spec
+cacheModelsSpec =
+  Hspec.around withTempDir . Hspec.describe "cacheAndUseModel" $ do
+    Hspec.it "caches a model" $ \fp -> do
+      cacheAndUseModel "mnist.ts.pt" . mkCacheOption fp $ 10 * 1073741824
+      (`Hspec.shouldBe` ["mnist.ts.pt"]) =<< listDirectory fp
```

Comment:
In my `inferno-ml` PR, I was trying to avoid adding the ~4MB mnist model to the git repo. Perhaps I could modify the GitHub action to move this model file that it downloads to the right location for these tests?

Reply:
I understand the desire behind that, but that means that this whole test won't be buildable with Nix (unless I reorganized it) and I'm also using it in some `nixosTest`s that require no network access (it needs to be pure)

Reply:
I see. Is it okay to add hasktorch as a dependency? Then I can provide a Haskell function that creates a random ML model of whatever size you require and save it as a `.ts.pt` file. Would that work?

Reply:
@ngua maybe you missed this comment?